Bluetooth: Classic: AVCTP browsing channel support

makeshi opened the pull request:
This patch introduces AVCTP Browsing channel support over L2CAP BR/EDR,
including implementation of L2CAP channel callbacks and browsing
operation APIs, and also add Kconfig for BT_AVCTP_BROWSING.

makeshi committed:
Bluetooth: AVCTP: Refactors the AVCTP to support multiple L2CAP PSMs

- This patch refactors the AVCTP layer to support multiple L2CAP PSMs
- Introduced bt_avctp_server structure for dynamic L2CAP server
  registration and new server register apis
- Added k_sem protected server list to support multiple AVCTP servers

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi committed:
bluetooth: avrcp: Add support for AVRCP browsing

- AVRCP SDP record updates for browsing channel
- L2CAP setup for AVCTP browsing channel, and add connect/disconnect
- Add SetBrowsedPlayer browsing command handling

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -255,3 +257,245 @@ int bt_avctp_init(void)
+	LOG_DBG("conn %p session %p", conn, session);
+
+	init_avctp_browsing_channel(session);
+	session->control_session = control_session;
```
Yes, thanks, will remove it in code and  add some comments is header file for AVCTP Browsing channel is dependent on the control session.

makeshi commented on subsys/bluetooth/host/classic/avctp_internal.h:
```diff
@@ -72,22 +73,39 @@ struct bt_avctp_header {
+struct bt_avctp_browsing {
+	struct bt_l2cap_br_chan br_chan;
+	const struct bt_avctp_browsing_ops_cb *ops;
+	struct bt_avctp *control_session;
```
It should be removed, thanks.

gzh-terry commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -255,3 +257,258 @@ int bt_avctp_init(void)
+
+	LOG_DBG("conn %p", conn);
+
+	if (!browsing_event_cb) {
```
Consider make it a macro as [AVRCP#85917](https://github.com/zephyrproject-rtos/zephyr/pull/85917/files#diff-a44f7831ad5cd74f4097b158581321890ff3bfc1d5e6e37bf087ae719d2da46eR60)

gzh-terry commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -255,3 +260,254 @@ int bt_avctp_init(void)
+	}
+
+#if defined(CONFIG_BT_AVRCP)
+	if (hdr->pid == sys_cpu_to_be16(BT_SDP_AV_REMOTE_SVCLASS)) {
```
`if (sys_be16_to_cpu(hdr->pid) == BT_SDP_AV_REMOTE_SVCLASS)`
would be more readable.
Use same conversion as 
```
	LOG_ERR("unsupported AVCTP browsing PID received: 0x%04x",
		sys_be16_to_cpu(hdr->pid));
```

lylezhu2012 reviewed (changes requested):
I do not think the `browsing channel` can be added into `AVCTP`. The control channel and browsing channel are the concept of `AVRCP`.

makeshi committed:
Bluetooth: Shell: AVRCP: Shell interface for browsing support

- Add shell interface for browsing connect and disconnect testing and
  SetBrowsedPlayer command testing.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -104,32 +109,28 @@ static int avctp_l2cap_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+	if (sys_be16_to_cpu(hdr->pid) == BT_SDP_AV_REMOTE_SVCLASS) {
+		return session->ops->recv(session, buf);
+	}
 #endif
```
Removed the CONFIG_BT_AVRCP checking, it better not checking here because it is a upper layer MACRO,

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -90,46 +90,32 @@ static int avctp_l2cap_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+		err = bt_avctp_send(session, rsp);
+		if (err < 0) {
+			net_buf_unref(rsp);
+			LOG_ERR("AVCTP send fail, err = %d", err);
```
I think we need to consider dropping the session in the failure case.

makeshi replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -195,63 +192,76 @@ int bt_avctp_send(struct bt_avctp *session, struct net_buf *buf)
+
+	k_mutex_lock(&avctp_server_lock, K_FOREVER);
+
+	if (!bt_avctp_find_l2cap_server(&avctp_l2cap_server, l2cap_server)) {
```
I think there is a function `sys_slist_find()` provided by kernel can be used here.

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -195,63 +192,76 @@ int bt_avctp_send(struct bt_avctp *session, struct net_buf *buf)
+
+	if (!bt_avctp_find_l2cap_server(&avctp_l2cap_server, l2cap_server)) {
+		LOG_WRN("Invalid l2cap server");
+		return -EINVAL;
```
Mutex unlock is missing.

makeshi replied:
fixed

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -195,63 +192,76 @@ int bt_avctp_send(struct bt_avctp *session, struct net_buf *buf)
+
+	k_mutex_lock(&avctp_server_lock, K_FOREVER);
+
+	if (bt_avctp_find_l2cap_server(&avctp_l2cap_server, server)) {
```
The function `sys_slist_find()` can be used here to replace `bt_avctp_find_l2cap_server()`.

makeshi replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp_internal.h:
```diff
@@ -82,6 +80,53 @@ struct bt_avctp_ops_cb {
+	 *
+	 * The `sec_level` field specifies the minimum required security level.
+	 *
+	 * Note: The `accept` callback in this structure is not used by GOEP applications.
```
Comments issue.

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp_internal.h:
```diff
@@ -82,6 +80,53 @@ struct bt_avctp_ops_cb {
+ * @brief AVCTP L2CAP Server structure
+ *
+ * This structure defines the L2CAP server used for AVCTP over L2CAP transport.
+ * It is typically used in GOEP v2.0 and later implementations.
```
Ditto

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -129,7 +174,11 @@ static struct bt_sdp_attribute avrcp_tg_attrs[] = {
 		)
 	),
-	BT_SDP_SUPPORTED_FEATURES(AVRCP_CAT_1 | AVRCP_CAT_2),
+	BT_SDP_SUPPORTED_FEATURES(AVRCP_CAT_1 | AVRCP_CAT_2
+#if defined(CONFIG_BT_AVRCP_BROWSING)
+				 |AVRCP_BROWSING_SUPPORT
+#endif /* CONFIG_BT_AVRCP_BROWSING */
+	),
```
My suggestion is,
Define a macro AVRCP_BROWSING_ENABLE,
```
#if defined(CONFIG_BT_AVRCP_BROWSING)
#define AVRCP_BROWSING_ENABLE AVRCP_BROWSING_SUPPORT
#else
#define AVRCP_BROWSING_ENABLE 0
#endif /* CONFIG_BT_AVRCP_BROWSING */
```

```suggestion
	BT_SDP_SUPPORTED_FEATURES(AVRCP_CAT_1 | AVRCP_CAT_2 | AVRCP_BROWSING_ENABLE),
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -186,28 +235,52 @@ static struct bt_sdp_attribute avrcp_ct_attrs[] = {
-	BT_SDP_SUPPORTED_FEATURES(AVRCP_CAT_1 | AVRCP_CAT_2),
-	/* O: Provider Name not presented */
+#endif /* CONFIG_BT_AVRCP_BROWSING */
+	BT_SDP_SUPPORTED_FEATURES(AVRCP_CAT_1 | AVRCP_CAT_2
+#if defined(CONFIG_BT_AVRCP_BROWSING)
+		| AVRCP_BROWSING_SUPPORT
+#endif /* CONFIG_BT_AVRCP_BROWSING */
```
Ditto

makeshi replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -729,20 +861,190 @@ static int avrcp_accept(struct bt_conn *conn, struct bt_avctp **session)
+		avrcp_ct_cb->browsing_connected(session->br_chan.chan.conn, get_avrcp_ct(avrcp));
+	}
+
+	if ((avrcp_tg_cb != NULL) && (avrcp_tg_cb->browsing_connected != NULL)) {
+		avrcp_tg_cb->browsing_connected(session->br_chan.chan.conn, get_avrcp_tg(avrcp));
+	}
```
Duplicated code block.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -729,20 +861,190 @@ static int avrcp_accept(struct bt_conn *conn, struct bt_avctp **session)
+	if ((avrcp_ct_cb != NULL) && (avrcp_ct_cb->disconnected != NULL)) {
+		avrcp_ct_cb->browsing_disconnected(get_avrcp_ct(avrcp));
+	}
+	if ((avrcp_tg_cb != NULL) && (avrcp_tg_cb->disconnected != NULL)) {
+		avrcp_tg_cb->browsing_disconnected(get_avrcp_tg(avrcp));
+	}
```
Ditto

makeshi replied:
It is not duplicated code block, this one for TG, above code for CT

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -729,20 +861,190 @@ static int avrcp_accept(struct bt_conn *conn, struct bt_avctp **session)
+	}
+
+	avctp_hdr = net_buf_pull_mem(buf, sizeof(*avctp_hdr));
+	if (buf->len < sizeof(struct bt_avrcp_avc_brow_pdu)) {
+		LOG_ERR("invalid AVRCP browsing header received");
+		return -EINVAL;
+	}
```
It is duplicated with lines 962 - 965.

makeshi replied:
thanks, removed

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -729,20 +861,190 @@ static int avrcp_accept(struct bt_conn *conn, struct bt_avctp **session)
+		return -EMSGSIZE;
+	}
+
+	avctp_hdr = net_buf_pull_mem(buf, sizeof(*avctp_hdr));
```
I think the checking of packet type is necessary.

makeshi replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -729,20 +861,190 @@ static int avrcp_accept(struct bt_conn *conn, struct bt_avctp **session)
+		return -ENOTCONN;
+	}
+
+	init_avctp_browsing_channel(&(avrcp->browsing_session));
```
I think we need to consider that the session `avrcp->browsing_session` has been connected.

makeshi replied:
updated, have added browsing_session_connected for check browsing_session has been connected.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -781,6 +1083,7 @@ int bt_avrcp_connect(struct bt_conn *conn)
 	}
 
 	avrcp->session.ops = &avctp_ops;
+	init_avctp_control_channel(&(avrcp->session));
```
Ditto

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -814,6 +1117,48 @@ int bt_avrcp_disconnect(struct bt_conn *conn)
+		return -ENOTCONN;
+	}
+
+	avrcp->browsing_session.ops = &browsing_avctp_ops;
```
Ditto

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -141,6 +141,33 @@ typedef enum __packed {
 	BT_AVRCP_BUTTON_RELEASED = 1,
 } bt_avrcp_button_state_t;
 
+
```
updated

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -977,3 +1348,67 @@ int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+		return -ENOMEM;
+	}
+
+	hdr = net_buf_add(buf, sizeof(struct bt_avrcp_avc_brow_pdu));
```
updated

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -977,3 +1348,67 @@ int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+	hdr->param_len = sys_cpu_to_be16(param_len);
+
+	/* Add status */
+	net_buf_add_u8(buf, rsp->status);
```
updated, checked by if (net_buf_tailroom(buf) < param_len) {

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -977,3 +1348,67 @@ int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+	net_buf_add_u8(buf, rsp->status);
+
+	if (rsp->status == BT_AVRCP_STATUS_OPERATION_COMPLETED) {
+		net_buf_add_be16(buf, rsp->uid_counter);
+		net_buf_add_be32(buf, rsp->num_items);
+		net_buf_add_be16(buf, rsp->charset_id);
+		net_buf_add_u8(buf, rsp->folder_depth);
```
updated, checked by if (net_buf_tailroom(buf) < param_len) {

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -977,3 +1348,67 @@ int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+		folder_ptr = (struct bt_avrcp_folder_name *)rsp->folder_names;
+		for (size_t i = 0; i < rsp->folder_depth; i++) {
+			if (folder_ptr->folder_name_len > 0U) {
+				net_buf_add_be16(buf, folder_ptr->folder_name_len);
```
updated, checked by if (net_buf_tailroom(buf) < param_len) {

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -264,6 +341,27 @@ int bt_avrcp_connect(struct bt_conn *conn);
  */
 int bt_avrcp_disconnect(struct bt_conn *conn);
```
have added the control channel checking in c file

makeshi commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -364,6 +463,111 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+	rsp.uid_counter = uid_counter;
+	rsp.num_items = num_items;
+	rsp.charset_id = charset_id;
+	rsp.folder_depth = folder_depth;
```
Updated: folder_depth is now set to 1 by default and removed from input arguments to avoid potential issues

makeshi commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -373,12 +577,18 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
+	SHELL_CMD_ARG(send_browsed_player_rsp, NULL, "Send SetBrowsedPlayer response\n"
+		      "Usage: send_browsed_player_rsp <status> <uid_counter> <num_items> "
+		      "<charset_id> <folder_depth> <folder_name>", cmd_send_set_browsed_player_rsp,
+		      1, 6),
```
fixed

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -814,6 +1124,53 @@ int bt_avrcp_disconnect(struct bt_conn *conn)
+	if (err < 0) {
+		LOG_DBG("AVCTP browsing connect failed");
+	}
+	avrcp->browsing_session_connected = true;
```
fixed

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -729,20 +854,204 @@ static int avrcp_accept(struct bt_conn *conn, struct bt_avctp **session)
+	LOG_DBG("browsing_session: %p", &(avrcp->browsing_session));
+
+	return 0;
+}
```
updated

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -814,6 +1124,53 @@ int bt_avrcp_disconnect(struct bt_conn *conn)
+
+	LOG_DBG("Browsing connection request sent");
+	return err;
+}
```
updated

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -31,8 +31,9 @@
-static const struct bt_avctp_event_cb *event_cb;
+/* L2CAP Server list */
+static sys_slist_t avctp_l2cap_server = SYS_SLIST_STATIC_INIT(&avctp_l2cap_server);
+static struct k_mutex avctp_server_lock;
```
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -814,6 +1134,59 @@ int bt_avrcp_disconnect(struct bt_conn *conn)
+	avrcp->browsing_session.ops = &browsing_avctp_ops;
+	init_avctp_browsing_channel(&(avrcp->browsing_session));
+	err = bt_avctp_connect(conn, &(avrcp->browsing_session));
+	if (err < 0) {
+		LOG_DBG("AVCTP browsing connect failed");
+	} else {
+		avrcp->browsing_session_connected = true;
+	}
+
+	LOG_DBG("Browsing connection request sent");
+	return err;
```
if (err < 0) {
		LOG_DBG("AVCTP browsing connect failed");
		return err;
	}

	avrcp->browsing_session_connected = true;
	LOG_DBG("Browsing connection request sent");

	return 0;

makeshi replied:
updated

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -239,6 +305,17 @@ struct bt_avrcp_ct_cb {
 	 */
 	void (*passthrough_rsp)(struct bt_avrcp_ct *ct, uint8_t tid, bt_avrcp_rsp_t result,
 				const struct bt_avrcp_passthrough_rsp *rsp);
+
+	/** @brief Callback function for bt_avrcp_ct_set_browsed_player().
+	 *
+	 *  Called when the set browsed player process is completed.
+	 *
+	 *  @param ct AVRCP CT connection object.
+	 *  @param tid The transaction label of the response.
+	 *  @param rsp The response for SET BROWSED PLAYER command.
+	 */
+	void (*browsed_player_rsp)(struct bt_avrcp_ct *ct, uint8_t tid,
+				   struct bt_avrcp_set_browsed_player_rsp *rsp);
```
My suggestion is passing the net buffer directly instead of self-defined structure. And provide the spec-defined structure with the PDU format. Then the net buffer can be parsed using the defined structures in application layer.

There are several advantages,
-  Reduce the stack usage,
-  Remove the memory copy and improve the performance,
-  Avoid the self-defined structure and use spec-defined structure. It can increase readability.

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -378,6 +518,20 @@ int bt_avrcp_tg_register_cb(const struct bt_avrcp_tg_cb *cb);
+ *
+ *  @return 0 in case of success or error code in case of error.
+ */
+int bt_avrcp_tg_send_set_browsed_player_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+					    const struct bt_avrcp_set_browsed_player_rsp *rsp);
```
Ditto.

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -354,6 +464,36 @@ struct bt_avrcp_tg_cb {
+	 *  @param req The set browsed player request data.
+	 */
+	void (*set_browsed_player_req)(struct bt_avrcp_tg *tg, uint8_t tid,
+				       const struct bt_avrcp_set_browsed_player_req *req);
```
```suggestion
				       uint16_t player_id);
```

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -90,46 +90,33 @@ static int avctp_l2cap_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+					  BT_AVCTP_PKT_TYPE_SINGLE, BT_AVCTP_IPID_INVALID,
+					  tid, hdr->pid);
+		if (!rsp) {
+			return -ENOMEM;
```
updated

makeshi commented on subsys/bluetooth/host/classic/avctp_internal.h:
```diff
@@ -82,6 +80,52 @@ struct bt_avctp_ops_cb {
 	struct bt_l2cap_br_chan br_chan;
 	const struct bt_avctp_ops_cb *ops;
+	uint16_t pid;  /** Profile Identifier */
+	uint16_t psm;
```
The PSMs for AVRCP Control and Browsing are different, so the upper-layer software needs to pass the correct PSM to avctp

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -141,6 +141,32 @@ typedef enum __packed {
+	BT_AVRCP_STATUS_PLAYER_NOT_ADDRESSED        = 0x13,
+	BT_AVRCP_STATUS_NO_VALID_SEARCH_RESULTS     = 0x14,
+	BT_AVRCP_STATUS_NO_AVAILABLE_PLAYERS        = 0x15,
+	BT_AVRCP_STATUS_ADDRESSED_PLAYER_CHANGED    = 0x16
```
updated

makeshi committed:
bluetooth: avrcp: provide an API for net buffer allocation

Add bt_avrcp_create_pdu() to allocate net_buf with protocol headroom.

Signed-off-by: Make Shi <make.shi@nxp.com>

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -729,21 +852,234 @@ static int avrcp_accept(struct bt_conn *conn, struct bt_avctp **session)
+		return -EINVAL;
+	}
+
+	if (buf->len < (sizeof(struct bt_avrcp_set_browsed_player_rsp) -
+			sizeof(rsp.folder_names))) {
+		LOG_ERR("Invalid AVRCP browsing header received: buffer too short (%u)", buf->len);
+		return -EMSGSIZE;
+	}
```
Similar comments here. If the `min_len` is added to `struct avrcp_pdu_handler`. The `buf->len` can be checked in common code.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -729,21 +852,234 @@ static int avrcp_accept(struct bt_conn *conn, struct bt_avctp **session)
+		goto error_rsp;
+	}
+
+	if (buf->len < sizeof(uint16_t)) {
+		LOG_ERR("Invalid set browsed player request length");
+		goto error_rsp;
+	}
```
Ditto,

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -115,13 +129,42 @@ static void avrcp_passthrough_rsp(struct bt_avrcp_ct *ct, uint8_t tid, bt_avrcp_
+	uint8_t status;
+	uint8_t folder_depth;
+	struct bt_avrcp_folder_name *folder_names;
+
```
Need to check `buf->len` before pulling data from `buf`.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -364,6 +474,122 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+		num_items = 100U;
+	}
+
+	if (argc >= 5) {
+		charset_id = (uint16_t)strtoul(argv[4], NULL, 0);
+	} else {
+		charset_id = 0x006AU; /* UTF-8 */
+	}
```
I think the charset should not be a configurable field when argc is 5, because the value of `folder_name` ("Music") is `UTF-8` format.

Or the argv[4] and argv[5] should be set at same time.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -364,6 +474,122 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+		charset_id = 0x006AU; /* UTF-8 */
+	}
+
+	if (argc >= 6) {
+		folder_name = argv[5];
+	}
+
+	folder_name_len = strlen(folder_name);
```
The `argv[5]` should be a hex string. Because when the charset is not `UTF-8`, the `argv[5]` cannot be treated as a string. It should be a hex string, and should be converted by `bin2hex` to a byte[].

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -977,3 +1428,41 @@ int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+	}
+
+	param_len = buf->len;
+	hdr = net_buf_push(buf, sizeof(struct bt_avrcp_avc_brow_pdu));
```
The headroom of buf should be checked before using it.

makeshi replied:
fixed

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -141,6 +141,32 @@ typedef enum __packed {
+	BT_AVRCP_STATUS_NO_VALID_SEARCH_RESULTS     = 0x14,
+	BT_AVRCP_STATUS_NO_AVAILABLE_PLAYERS        = 0x15,
+	BT_AVRCP_STATUS_ADDRESSED_PLAYER_CHANGED    = 0x16,
+} bt_avrcp_browsing_status_t;
```
May change this type name as well.

makeshi replied:
changed

gzh-terry commented on subsys/bluetooth/host/classic/avrcp_internal.h:
```diff
@@ -12,17 +12,32 @@
+#define AVRCP_BROWSING_SUPPORT            BIT(6)
+#define AVRCP_MULTIPLE_MEDIA_PLAYERS      BIT(7)
+#define AVRCP_COVER_ART_SUPPORT           BIT(8)
+#define AVRCP_RFA_MASK                    0xFF00  /* Bits 9-15 reserved for future */
```
Looks not necessary to add this macro.

makeshi replied:
removed

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -141,6 +141,32 @@ typedef enum __packed {
 	BT_AVRCP_BUTTON_RELEASED = 1,
 } bt_avrcp_button_state_t;
 
+/* AVRCP Browsing Status Codes */
```
thanks, updated

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -141,6 +141,32 @@ typedef enum __packed {
+typedef enum __packed {
+	BT_AVRCP_STATUS_INVALID_COMMAND             = 0x00,
+	BT_AVRCP_STATUS_INVALID_PARAMETER           = 0x01,
+	BT_AVRCP_STATUS_PARAMETER_NOT_FOUND         = 0x02,
```
updated

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -141,6 +141,32 @@ typedef enum __packed {
+	BT_AVRCP_STATUS_DOES_NOT_EXIST              = 0x09,
+	BT_AVRCP_STATUS_INVALID_SCOPE               = 0x0A,
+	BT_AVRCP_STATUS_RANGE_OUT_OF_BOUNDS         = 0x0B,
+	BT_AVRCP_STATUS_UID_IS_A_DIRECTORY          = 0x0C,
```
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -805,6 +1141,15 @@ int bt_avrcp_disconnect(struct bt_conn *conn)
+		err = bt_avrcp_browsing_disconnect(conn);
+		if (err) {
+			LOG_ERR("Browsing session disconnect failed: %d", err);
+			return -EIO;
```
```suggestion
			return err;
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -814,6 +1159,69 @@ int bt_avrcp_disconnect(struct bt_conn *conn)
+
+	err = bt_avctp_disconnect(&(avrcp->browsing_session));
+	if (err < 0) {
+		LOG_DBG("AVCTP browsing disconnect failed");
```
```suggestion
		LOG_ERR("AVCTP browsing disconnect failed");
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -814,6 +1159,69 @@ int bt_avrcp_disconnect(struct bt_conn *conn)
+	init_avctp_browsing_channel(&(avrcp->browsing_session));
+	err = bt_avctp_connect(conn, BT_L2CAP_PSM_AVRCP_BROWSING, &(avrcp->browsing_session));
+	if (err < 0) {
+		LOG_DBG("AVCTP browsing connect failed");
```
```suggestion
		LOG_ERR("AVCTP browsing connect failed");
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -919,6 +1327,51 @@ int bt_avrcp_ct_passthrough(struct bt_avrcp_ct *ct, uint8_t tid, uint8_t opid, u
+	}
+
+	buf = avrcp_create_browsing_pdu(ct->avrcp, tid, BT_AVCTP_CMD);
+	if (!buf) {
```
```suggestion
	if (buf == NULL) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -919,6 +1327,51 @@ int bt_avrcp_ct_passthrough(struct bt_avrcp_ct *ct, uint8_t tid, uint8_t opid, u
+	net_buf_add_be16(buf, player_id);
+
+	err =  avrcp_browsing_send(ct->avrcp, buf);
+	if (err) {
```
```suggestion
	if (err < 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -977,3 +1430,57 @@ int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+			    sys_cpu_to_be16(BT_SDP_AV_REMOTE_SVCLASS));
+
+	err =  avrcp_browsing_send(tg->avrcp, buf);
+	if (err) {
```
```suggestion
	if (err < 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -256,6 +347,53 @@ static int cmd_disconnect(const struct shell *sh, int32_t argc, char *argv[])
+	}
+
+	err = bt_avrcp_browsing_connect(default_conn);
+	if (err) {
```
```suggestion
	if (err < 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -256,6 +347,53 @@ static int cmd_disconnect(const struct shell *sh, int32_t argc, char *argv[])
+
+	if (default_ct != NULL) {
+		err = bt_avrcp_browsing_disconnect(default_conn);
+		if (err) {
```
Ditto.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -364,6 +502,137 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+	player_id = (uint16_t)strtoul(argv[1], NULL, 0);
+
+	err = bt_avrcp_ct_set_browsed_player(default_ct, get_next_tid(), player_id);
+	if (err) {
```
Ditto.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -364,6 +502,137 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+{
+	struct bt_avrcp_set_browsed_player_rsp *rsp;
+	char *folder_name_str = "Music";
+	uint8_t folder_name_hex[80];
```
It is better to define a macro for size of `folder_name_hex` instead of hard code.

makeshi replied:
have defined FOLDER_NAME_HEX_BUF_LEN

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -364,6 +502,137 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+		}
+	}
+
+	param_len = folder_name_len + sizeof(uint16_t);
```
```suggestion
	param_len = folder_name_len + sizeof(struct bt_avrcp_folder_name);
```

makeshi replied:
done

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -364,6 +502,137 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+		return -ENOMEM;
+	}
+
+	net_buf_add_be16(buf, folder_name_len);
```
```suggestion
	struct bt_avrcp_folder_name *folder_name;

	folder_name = net_buf_add(buf, sizeof(*folder_name));
	folder_name->folder_name_len = sys_cput_to_be16(folder_name_len);

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -364,6 +502,137 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+	}
+
+	err = bt_avrcp_tg_send_set_browsed_player_rsp(default_tg, tg_tid, buf);
+	if (!err) {
```
```suggestion
	if (err == 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -364,6 +502,137 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+
+	if (net_buf_tailroom(buf) < sizeof(struct bt_avrcp_set_browsed_player_rsp)) {
+		shell_error(sh, "Not enough tailroom in buffer for browsed player rsp");
+		net_buf_unref(buf);
+		return -ENOMEM;
```
The suggested error code of shell handler is `-ENOEXEC`.

It is recommended to use `goto` for all cases that the error is not 0 and the `buf` needs to be released.
```suggestion
		goto failed;
```
And add `failed` at the end of the function. Such as
```
failed:
	net_buf_unref(buf);
	return -ENOEXEC;
```

makeshi replied:
updated
